Index: linux/drivers/net/wireless/rockchip_wlan/rkwifi/bcmdhd/wl_cfg80211.c
===================================================================
--- linux.orig/drivers/net/wireless/rockchip_wlan/rkwifi/bcmdhd/wl_cfg80211.c
+++ linux/drivers/net/wireless/rockchip_wlan/rkwifi/bcmdhd/wl_cfg80211.c
@@ -15176,7 +15176,7 @@ wl_cfg80211_netdev_notifier_call(struct
 
 	wdev = ndev_to_wdev(dev);
 	if (!wdev) {
-		WL_ERR(("wdev null. Do nothing\n"));
+		WL_DBG(("wdev null. Do nothing\n"));
 		return NOTIFY_DONE;
 	}
 
