Index: linux/arch/arm64/boot/dts/rockchip/rk3588-evb1-lp4.dtsi
===================================================================
--- linux.orig/arch/arm64/boot/dts/rockchip/rk3588-evb1-lp4.dtsi
+++ linux/arch/arm64/boot/dts/rockchip/rk3588-evb1-lp4.dtsi
@@ -304,6 +304,8 @@
 	reset-gpios = <&gpio2 RK_PB4 GPIO_ACTIVE_LOW>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&lcd_rst_gpio>;
+  width-mm = <68>;
+  height-mm = <120>;
 };
 
 /*
