Index: linux/drivers/usb/typec/tcpm/tcpm.c
===================================================================
--- linux.orig/drivers/usb/typec/tcpm/tcpm.c
+++ linux/drivers/usb/typec/tcpm/tcpm.c
@@ -6443,7 +6443,7 @@ static int devm_tcpm_psy_register(struct
 	snprintf(psy_name, psy_name_len, "%s%s", tcpm_psy_name_prefix,
 		 port_dev_name);
 	port->psy_desc.name = psy_name;
-	port->psy_desc.type = POWER_SUPPLY_TYPE_USB,
+	port->psy_desc.type = POWER_SUPPLY_TYPE_USB_TYPE_C,
 	port->psy_desc.usb_types = tcpm_psy_usb_types;
 	port->psy_desc.num_usb_types = ARRAY_SIZE(tcpm_psy_usb_types);
 	port->psy_desc.properties = tcpm_psy_props,
