Index: linux/drivers/input/touchscreen/focaltech_touch/focaltech_core.c
===================================================================
--- linux.orig/drivers/input/touchscreen/focaltech_touch/focaltech_core.c
+++ linux/drivers/input/touchscreen/focaltech_touch/focaltech_core.c
@@ -1781,3 +1781,4 @@ module_exit(fts_ts_exit);
 MODULE_AUTHOR("FocalTech Driver Team");
 MODULE_DESCRIPTION("FocalTech Touchscreen Driver");
 MODULE_LICENSE("GPL v2");
+MODULE_IMPORT_NS(VFS_internal_I_am_really_a_filesystem_and_am_NOT_a_driver);
