Index: linux/arch/arm64/boot/dts/rockchip/rk3588s-orangepi.dtsi
===================================================================
--- linux.orig/arch/arm64/boot/dts/rockchip/rk3588s-orangepi.dtsi
+++ linux/arch/arm64/boot/dts/rockchip/rk3588s-orangepi.dtsi
@@ -562,6 +562,7 @@
 
 &vop {
 	status = "okay";
+    disable-win-move;
 	assigned-clocks = <&cru ACLK_VOP>;
 	assigned-clock-rates = <800000000>;
 };
@@ -576,23 +577,27 @@
 
 /* vp0 & vp1 splice for 8K output */
 &vp0 {
+  cursor-win-id=<ROCKCHIP_VOP2_ESMART0>;
 	rockchip,plane-mask = <(1 << ROCKCHIP_VOP2_CLUSTER0 | 1 << ROCKCHIP_VOP2_ESMART0)>;
-	rockchip,primary-plane = <ROCKCHIP_VOP2_ESMART0>;
+	rockchip,primary-plane = <ROCKCHIP_VOP2_CLUSTER0>;
 };
 
 &vp1 {
+  cursor-win-id=<ROCKCHIP_VOP2_ESMART1>;
 	rockchip,plane-mask = <(1 << ROCKCHIP_VOP2_CLUSTER1 | 1 << ROCKCHIP_VOP2_ESMART1)>;
-	rockchip,primary-plane = <ROCKCHIP_VOP2_ESMART1>;
+	rockchip,primary-plane = <ROCKCHIP_VOP2_CLUSTER1>;
 };
 
 &vp2 {
+  cursor-win-id=<ROCKCHIP_VOP2_ESMART2>;
 	rockchip,plane-mask = <(1 << ROCKCHIP_VOP2_CLUSTER2 | 1 << ROCKCHIP_VOP2_ESMART2)>;
-	rockchip,primary-plane = <ROCKCHIP_VOP2_ESMART2>;
+	rockchip,primary-plane = <ROCKCHIP_VOP2_CLUSTER2>;
 };
 
 &vp3 {
+  cursor-win-id=<ROCKCHIP_VOP2_ESMART3>;
 	rockchip,plane-mask = <(1 << ROCKCHIP_VOP2_CLUSTER3 | 1 << ROCKCHIP_VOP2_ESMART3)>;
-	rockchip,primary-plane = <ROCKCHIP_VOP2_ESMART3>;
+	rockchip,primary-plane = <ROCKCHIP_VOP2_CLUSTER3>;
 };
 
 /* Fix tty terminal out of screen, and most dclk of resolutions was not supported in hdmiphy clock from parent clock by default */
