diff --git a/drivers/net/wireless/rockchip_wlan/rtl8852be/Makefile b/drivers/net/wireless/rockchip_wlan/rtl8852be/Makefile
index 5a7fb242e403..6f5a3168b984 100755
--- a/drivers/net/wireless/rockchip_wlan/rtl8852be/Makefile
+++ b/drivers/net/wireless/rockchip_wlan/rtl8852be/Makefile
@@ -20,7 +20,8 @@ ifeq ($(GCC_VER_49),1)
 EXTRA_CFLAGS += -Wno-date-time	# Fix compile error && warning on gcc 4.9 and later
 endif
 
-EXTRA_CFLAGS += -I$(src)/include
+#EXTRA_CFLAGS += -I$(src)/include
+EXTRA_CFLAGS += -I$(S)/drivers/net/wireless/rockchip_wlan/rtl8852be/include
 
 EXTRA_LDFLAGS += --strip-debug
 
@@ -199,6 +200,7 @@ DRV_PATH = $(src)
 else
 DRV_PATH = $(TopDIR)
 endif
+DRV_PATH = $(S)/drivers/net/wireless/rockchip_wlan/rtl8852be
 
 ########### HAL_RTL8852A #################################
 ifeq ($(CONFIG_RTL8852A), y)
@@ -584,7 +586,9 @@ endif
 include $(wildcard $(DRV_PATH)/platform/*.mk)
 
 # Import platform specific compile options
-EXTRA_CFLAGS += -I$(src)/platform
+#EXTRA_CFLAGS += -I$(src)/platform
+EXTRA_CFLAGS += -I$(S)/drivers/net/wireless/rockchip_wlan/rtl8852be/platform
+
 #_PLATFORM_FILES := platform/platform_ops.o
 OBJS += $(_PLATFORM_FILES)
 
@@ -596,7 +600,8 @@ endif
 
 ifneq ($(KERNELRELEASE),)
 ########### COMMON #################################
-include $(src)/common.mk
+#include $(src)/common.mk
+include $(S)/drivers/net/wireless/rockchip_wlan/rtl8852be/common.mk
 
 EXTRA_CFLAGS += -DPHL_PLATFORM_LINUX
 EXTRA_CFLAGS += -DCONFIG_PHL_ARCH
@@ -617,7 +622,8 @@ ifeq ($(DIRTY_FOR_WORK), y)
 EXTRA_CFLAGS += -DDIRTY_FOR_WORK
 endif
 
-include $(src)/phl/phl.mk
+#include $(src)/phl/phl.mk
+include $(S)/drivers/net/wireless/rockchip_wlan/rtl8852be/phl/phl.mk
 
 
 obj-$(CONFIG_RTL8852BE) := $(MODULE_NAME).o
diff --git a/drivers/net/wireless/rockchip_wlan/rtl8852be/common.mk b/drivers/net/wireless/rockchip_wlan/rtl8852be/common.mk
index a32ab269dcde..a045c2afb8ad 100755
--- a/drivers/net/wireless/rockchip_wlan/rtl8852be/common.mk
+++ b/drivers/net/wireless/rockchip_wlan/rtl8852be/common.mk
@@ -89,7 +89,8 @@ _CORE_FILES :=	core/rtw_cmd.o \
 _CORE_FILES +=	core/rtw_phl.o \
 		core/rtw_phl_cmd.o
 
-EXTRA_CFLAGS += -I$(src)/core/crypto
+#EXTRA_CFLAGS += -I$(src)/core/crypto
+EXTRA_CFLAGS += -I$(S)/drivers/net/wireless/rockchip_wlan/rtl8852be/core/crypto
 _CORE_FILES += core/crypto/aes-internal.o \
 		core/crypto/aes-internal-enc.o \
 		core/crypto/aes-gcm.o \
diff --git a/drivers/net/wireless/rockchip_wlan/rtl8852bu/Makefile b/drivers/net/wireless/rockchip_wlan/rtl8852bu/Makefile
index 518564a8134f..9fee78c00669 100755
--- a/drivers/net/wireless/rockchip_wlan/rtl8852bu/Makefile
+++ b/drivers/net/wireless/rockchip_wlan/rtl8852bu/Makefile
@@ -20,7 +20,8 @@ ifeq ($(GCC_VER_49),1)
 EXTRA_CFLAGS += -Wno-date-time	# Fix compile error && warning on gcc 4.9 and later
 endif
 
-EXTRA_CFLAGS += -I$(src)/include
+#EXTRA_CFLAGS += -I$(src)/include
+EXTRA_CFLAGS += -I$(S)/drivers/net/wireless/rockchip_wlan/rtl8852bu/include
 
 EXTRA_LDFLAGS += --strip-debug
 
@@ -198,6 +199,8 @@ else
 DRV_PATH = $(TopDIR)
 endif
 
+DRV_PATH = $(S)/drivers/net/wireless/rockchip_wlan/rtl8852bu
+
 ########### HAL_RTL8852A #################################
 ifeq ($(CONFIG_RTL8852A), y)
 IC_NAME := rtl8852a
@@ -585,7 +588,8 @@ endif
 include $(wildcard $(DRV_PATH)/platform/*.mk)
 
 # Import platform specific compile options
-EXTRA_CFLAGS += -I$(src)/platform
+#EXTRA_CFLAGS += -I$(src)/platform
+EXTRA_CFLAGS += -I$(S)/drivers/net/wireless/rockchip_wlan/rtl8852bu/platform
 #_PLATFORM_FILES := platform/platform_ops.o
 OBJS += $(_PLATFORM_FILES)
 
@@ -597,7 +601,8 @@ endif
 
 ifneq ($(KERNELRELEASE),)
 ########### COMMON #################################
-include $(src)/common.mk
+#include $(src)/common.mk
+include $(S)/drivers/net/wireless/rockchip_wlan/rtl8852bu/common.mk
 
 EXTRA_CFLAGS += -DPHL_PLATFORM_LINUX
 EXTRA_CFLAGS += -DCONFIG_PHL_ARCH
@@ -618,8 +623,8 @@ ifeq ($(DIRTY_FOR_WORK), y)
 EXTRA_CFLAGS += -DDIRTY_FOR_WORK
 endif
 
-include $(src)/phl/phl.mk
-
+#include $(src)/phl/phl.mk
+include $(S)/drivers/net/wireless/rockchip_wlan/rtl8852bu/phl/phl.mk
 
 obj-$(CONFIG_RTL8852BU) := $(MODULE_NAME).o
 obj-$(CPTCFG_RTL8852AE) := $(MODULE_NAME).o
diff --git a/drivers/net/wireless/rockchip_wlan/rtl8852bu/common.mk b/drivers/net/wireless/rockchip_wlan/rtl8852bu/common.mk
index a32ab269dcde..31b2b74de681 100755
--- a/drivers/net/wireless/rockchip_wlan/rtl8852bu/common.mk
+++ b/drivers/net/wireless/rockchip_wlan/rtl8852bu/common.mk
@@ -89,7 +89,8 @@ _CORE_FILES :=	core/rtw_cmd.o \
 _CORE_FILES +=	core/rtw_phl.o \
 		core/rtw_phl_cmd.o
 
-EXTRA_CFLAGS += -I$(src)/core/crypto
+#EXTRA_CFLAGS += -I$(src)/core/crypto
+EXTRA_CFLAGS += -I$(S)/drivers/net/wireless/rockchip_wlan/rtl8852bu/core/crypto
 _CORE_FILES += core/crypto/aes-internal.o \
 		core/crypto/aes-internal-enc.o \
 		core/crypto/aes-gcm.o \
