diff --git a/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi b/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
index a0bf346ee795..51986e3aacab 100644
--- a/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
@@ -87,7 +87,7 @@ backlight: backlight {
 	dp0_sound: dp0-sound {
 		status = "disabled";
 		compatible = "rockchip,hdmi";
-		rockchip,card-name= "rockchip,dp0";
+		rockchip,card-name= "dp0-HDMI";
 		rockchip,mclk-fs = <512>;
 		rockchip,cpu = <&spdif_tx2>;
 		rockchip,codec = <&dp0 1>;
@@ -96,7 +96,7 @@ dp0_sound: dp0-sound {
 
 	dp1_sound: dp1-sound {
 		status = "disabled";
-		compatible = "rockchip,hdmi";
+		compatible = "dp1-HDMI";
 		rockchip,card-name= "rockchip,dp1";
 		rockchip,mclk-fs = <512>;
 		rockchip,cpu = <&spdif_tx5>;
@@ -108,7 +108,7 @@ hdmi0_sound: hdmi0-sound {
 		status = "disabled";
 		compatible = "rockchip,hdmi";
 		rockchip,mclk-fs = <128>;
-		rockchip,card-name = "rockchip-hdmi0";
+		rockchip,card-name = "rockchip0-HDMI";
 		rockchip,cpu = <&i2s5_8ch>;
 		rockchip,codec = <&hdmi0>;
 		rockchip,jack-det;
@@ -118,7 +118,7 @@ hdmi1_sound: hdmi1-sound {
 		status = "disabled";
 		compatible = "rockchip,hdmi";
 		rockchip,mclk-fs = <128>;
-		rockchip,card-name = "rockchip-hdmi1";
+		rockchip,card-name = "rockchip1-HDMI";
 		rockchip,cpu = <&i2s6_8ch>;
 		rockchip,codec = <&hdmi1>;
 		rockchip,jack-det;
